File: backend/app/models/support.py
"""
Break-glass support access.

A platform administrator deliberately cannot read a client's payroll: entity
access is decided by organization membership, which never consults the platform
role. That is the right default — a standing developer login over every
client's salary, PAN and bank data is the most attractive single target in the
system — and it leaves no way in when a client reports a bug they cannot
reproduce for you.

This is that way in, built so that using it is never quiet.

**Time-boxed.** A grant carries an expiry and dies on its own. Nothing here
depends on someone remembering to close a session.

**Reason-required.** A free-text reason is mandatory and stored. "Investigating
ticket 412" is an answer; an empty box is not.

**Read-only and masked, always.** Not flags — columns, recorded per grant, so a
later change of policy cannot rewrite what a past session was allowed to see.
Almost no bug lives in an individual salary; they live in configuration,
findings and totals, which masking leaves entirely legible.

**Visible to the client.** Every grant, and every use of one, writes to the
organization's own audit trail. The client can revoke instantly, and can
require approval or refuse support access altogether.
"""
import uuid
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    String,
    Text,
    Uuid,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base

# What a client allows. Stored on the organization.
SUPPORT_POLICIES = ("break_glass", "approval_required", "disabled")

# "expired" is not stored — it is derived from ``expires_at``, so a grant that
# lapsed overnight does not need a sweeper to have run to stop working.
GRANT_STATES = ("pending", "active", "ended", "revoked")

DEFAULT_MINUTES = 60
MAX_MINUTES = 480  # eight hours; longer is a conversation, not an incident


class SupportAccessGrant(Base):
    """One episode of platform staff being able to read one client's data."""

    __tablename__ = "support_access_grants"
    __table_args__ = (
        Index("ix_support_grants_org_state", "org_id", "state"),
        Index("ix_support_grants_user_state", "admin_user_id", "state"),
    )

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    org_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    admin_user_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # Kept as text as well: the account may be deleted, and "who looked at our
    # payroll in March" must still have an answer afterwards.
    admin_email: Mapped[str | None] = mapped_column(String(255))

    reason: Mapped[str] = mapped_column(Text, nullable=False)
    state: Mapped[str] = mapped_column(String(16), nullable=False, default="active")

    # The policy in force when this grant was made, copied rather than joined.
    # A client who switches to approval_required next month has not changed what
    # this grant was.
    policy_at_grant: Mapped[str] = mapped_column(String(24), nullable=False, default="break_glass")

    # Both are always true today. Stored per grant so that if the product ever
    # gains a wider mode, the record of what this session could see stays true.
    read_only: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    identity_masked: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    requested_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    approved_by_email: Mapped[str | None] = mapped_column(String(255))
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    # Who closed it, and whether they were the client revoking or the engineer
    # finishing — different facts, and the client cares which.
    ended_by_email: Mapped[str | None] = mapped_column(String(255))
    ended_reason: Mapped[str | None] = mapped_column(String(255))

    # Bumped on each request that used this grant, so a session that was opened
    # and never used is distinguishable from one that read the whole book.
    last_used_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    use_count: Mapped[int] = mapped_column(nullable=False, default=0)

File: frontend/src/lib/support.ts
import { apiFetch, parseEnvelopeResponse } from "@/lib/api";

/**
 * Break-glass support access, from both sides.
 *
 * The client's side is the one that matters: they see every session, can revoke
 * one instantly, and own the policy that decides whether sessions can be opened
 * at all.
 */

export type SupportPolicy = "break_glass" | "approval_required" | "disabled";

export type SupportGrant = {
  id: string;
  org_id?: string;
  organization?: string;
  admin_email: string | null;
  reason: string;
  state: "pending" | "active" | "ended" | "revoked" | "expired";
  policy_at_grant?: SupportPolicy;
  read_only: boolean;
  identity_masked: boolean;
  requested_at?: string | null;
  expires_at: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  ended_at?: string | null;
  ended_by?: string | null;
  ended_reason?: string | null;
  last_used_at?: string | null;
  use_count?: number;
};

export type SupportStatus = {
  policy: SupportPolicy;
  policies: { key: SupportPolicy; label: string; hint: string }[];
  active: SupportGrant[];
  history: SupportGrant[];
  always_true: string[];
};

export type SupportOrg = {
  id: string;
  name: string;
  org_type: string;
  support_access_policy: SupportPolicy;
};

/** Any member can see whether someone is looking right now. */
export const fetchActiveSupport = () =>
  apiFetch("/api/org/support/active").then((r) =>
    parseEnvelopeResponse<{ active: SupportGrant[] }>(r),
  );

/** Owners and managers see the policy and the full history. */
export const fetchSupportStatus = () =>
  apiFetch("/api/org/support").then((r) => parseEnvelopeResponse<SupportStatus>(r));

export const setSupportPolicy = (policy: SupportPolicy) =>
  apiFetch("/api/org/support/policy", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ policy }),
  }).then((r) => parseEnvelopeResponse<{ policy: SupportPolicy }>(r));

export const revokeSupport = (id: string) =>
  apiFetch(`/api/org/support/grants/${id}/revoke`, { method: "POST" }).then((r) =>
    parseEnvelopeResponse<SupportGrant>(r),
  );

export const approveSupport = (id: string) =>
  apiFetch(`/api/org/support/grants/${id}/approve`, { method: "POST" }).then((r) =>
    parseEnvelopeResponse<SupportGrant>(r),
  );

// ---- platform side --------------------------------------------------------
export const fetchSupportOrgs = () =>
  apiFetch("/api/admin/support/organizations").then((r) =>
    parseEnvelopeResponse<SupportOrg[]>(r),
  );

export const fetchMyGrants = () =>
  apiFetch("/api/admin/support/grants?mine=true").then((r) =>
    parseEnvelopeResponse<SupportGrant[]>(r),
  );

export const openSupport = (body: { org_id: string; reason: string; minutes: number }) =>
  apiFetch("/api/admin/support/grants", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => parseEnvelopeResponse<SupportGrant>(r));

export const endSupport = (id: string, reason: string) =>
  apiFetch(`/api/admin/support/grants/${id}/end`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ reason }),
  }).then((r) => parseEnvelopeResponse<SupportGrant>(r));

/** How long is left, in words. Never negative — an elapsed session is over. */
export function timeLeft(expiresAt: string | null): string {
  if (!expiresAt) return "—";
  const ms = new Date(expiresAt).getTime() - Date.now();
  if (ms <= 0) return "expired";
  const minutes = Math.round(ms / 60000);
  if (minutes < 60) return `${minutes} min left`;
  const hours = Math.floor(minutes / 60);
  return `${hours}h ${minutes % 60}m left`;
}
